src/aiida_quantumespresso/common/hubbard.py

```diff
@@ -0,0 +1,161 @@
+# -*- coding: utf-8 -*-
+"""Utility class and functions for HubbardStructureData."""
+# pylint: disable=no-name-in-module, invalid-name
+from __future__ import annotations
+
+from typing import List, Literal, Tuple
+
+from pydantic import BaseModel, conint, constr, validator
+
+
+class HubbardParameters(BaseModel):
+    """Class for describing onsite and intersite Hubbard interaction parameters.
+
+    .. note: allowed manifold formats are:
+            * {N}{L} (2 characters)
+            * {N1}{L1}-{N2}{L2} (5 characters)
+
+        N = quantum number (1,2,3,...); L = orbital letter (s,p,d,f,g,h)
+    """
+
+    atom_index: conint(strict=True, ge=0)
+    """Atom index in the abstract structure."""
+
+    atom_manifold: constr(strip_whitespace=True, to_lower=True, min_length=2, max_length=5)
+    """Atom manifold (syntax is `3d`, `3d-2p`)."""
+
+    neighbour_index: conint(strict=True, ge=0)
+    """Neighbour index in the abstract structure."""
+
+    neighbour_manifold: constr(strip_whitespace=True, to_lower=True, min_length=2, max_length=5)
+    """Atom manifold (syntax is `3d`, `3d-2p`)."""
+
+    translation: Tuple[conint(strict=True), conint(strict=True), conint(strict=True)]
+    """Translation vector referring to the neighbour atom, (3,) shape list of ints."""
+
+    value: float
+    """Value of the Hubbard parameter, expessed in eV."""
+
+    hubbard_type: Literal['Ueff', 'U', 'V', 'J', 'B', 'E2', 'E3']
+    """Type of the Hubbard parameters used (`Ueff`, `U`, `V`, `J`, `B`, `E2`, `E3`)."""
+
+    @validator('atom_manifold', 'neighbour_manifold')  # cls is mandatory to use
+    def check_manifolds(cls, value):  # pylint: disable=no-self-argument, no-self-use
+        """Check the validity of the manifold input.
+
+        Allowed formats are:
+            * {N}{L} (2 characters)
+            * {N1}{L1}-{N2}{L2} (5 characters)
+
+        N = quantum number (1,2,3,...); L = orbital letter (s,p,d,f,g,h)
+        """
+        length = len(value)
+        if length not in [2, 5]:
+            raise ValueError(f'invalid length ``{length}``. Only 2 or 5.')
+        if length == 2:
+            if not value[0] in [str(_ + 1) for _ in range(6)]:
+                raise ValueError(f'invalid quantum number {value[0]}')
+            if not value[1] in ['s', 'p', 'd', 'f', 'h']:
+                raise ValueError(f'invalid manifold symbol {value[1]}')
+        if length == 5:
+            if not value[2] == '-':
+                raise ValueError(f'the separator {value[0]} is not allowed. Only `-`')
+            if not value[3] in [str(_ + 1) for _ in range(6)]:
+                raise ValueError(f'the quantum number {value[0]} is not correct')
+            if not value[4] in ['s', 'p', 'd', 'f', 'h']:
+                raise ValueError(f'the manifold number {value[1]} is not correct')
+        return value
+
+    def to_tuple(self) -> Tuple[int, str, int, str, float, Tuple[int, int, int], str]:
```

Comment:
Just FYI, with the `from __future__ import annotations` you can even just use the normal `tuple`, `list`, `dict` etc. for type hints
```suggestion
    def to_tuple(self) -> tuple[int, str, int, str, float, tuple[int, int, int], str]:
```